src/game/AI/ScriptDevAI/scripts/northrend/naxxramas/boss_anubrekhan.cpp

```diff
@@ -21,127 +21,105 @@ SDComment: Intro text usage is not very clear. Requires additional research.
+        m_isRegularMode(creature->GetMap()->IsRegularDifficulty())
     {
-        m_pInstance = (instance_naxxramas*)pCreature->GetInstanceData();
-        m_bIsRegularMode = pCreature->GetMap()->IsRegularDifficulty();
-        m_introDialogue.InitializeDialogueHelper(m_pInstance);
-        m_bHasTaunted = false;
+        m_hasTaunted = false;
+        SetDataType(TYPE_ANUB_REKHAN);
+        AddOnAggroText(SAY_AGGRO1, SAY_AGGRO2, SAY_AGGRO3);
+        AddOnKillText(SAY_SLAY);
+        AddCustomAction(ANUBREKHAN_SUMMON, true, [&](){
+            DoCastSpellIfCan(nullptr, SPELL_SUMMON_GUARD);
+        });
         Reset();
-
-        DoCastSpellIfCan(m_creature, SPELL_DOUBLE_ATTACK, CAST_TRIGGERED | CAST_AURA_NOT_PRESENT);
     }
 
-    instance_naxxramas* m_pInstance;
-    DialogueHelper m_introDialogue;
-    bool m_bIsRegularMode;
-
-    uint32 m_uiImpaleTimer;
-    uint32 m_uiLocustSwarmTimer;
-    uint32 m_uiSummonTimer;
-    bool   m_bHasTaunted;
+    instance_naxxramas* m_instance;
+    bool m_isRegularMode;
+    bool m_hasTaunted;
 
     void Reset() override
     {
-        m_uiImpaleTimer      = 15000;
-        m_uiLocustSwarmTimer = 90000;
-        m_uiSummonTimer      = m_bIsRegularMode ? 20000 : 0;// spawn a guardian only after 20 seconds in normal mode; in heroic there are already 2 Guards spawned
+        m_creature->SetSpellList(m_isRegularMode ? SPELLSET_10N : SPELLSET_25N);
+        DoCastSpellIfCan(m_creature, SPELL_DOUBLE_ATTACK, CAST_TRIGGERED | CAST_AURA_NOT_PRESENT);
+    }
+
+    void Aggro(Unit *who) override
+    {
+        if (m_isRegularMode)
+            ResetTimer(ANUBREKHAN_SUMMON, 20s);
+        BossAI::Aggro(who);
     }
 
     void KilledUnit(Unit* pVictim) override
     {
+        if (!pVictim)
+            return;
         // Force the player to spawn corpse scarabs via spell
         if (pVictim->GetTypeId() == TYPEID_PLAYER)
             pVictim->CastSpell(pVictim, SPELL_SELF_SPAWN_5, TRIGGERED_OLD_TRIGGERED, nullptr, nullptr, m_creature->GetObjectGuid());
 
-        if (urand(0, 4))
-            return;
-
-        DoScriptText(SAY_SLAY, m_creature);
-    }
-
-    void Aggro(Unit* /*pWho*/) override
-    {
-        switch (urand(0, 2))
-        {
-            case 0: DoScriptText(SAY_AGGRO1, m_creature); break;
-            case 1: DoScriptText(SAY_AGGRO2, m_creature); break;
-            case 2: DoScriptText(SAY_AGGRO3, m_creature); break;
-        }
-
-        if (m_pInstance)
-            m_pInstance->SetData(TYPE_ANUB_REKHAN, IN_PROGRESS);
-    }
-
-    void JustDied(Unit* /*pKiller*/) override
-    {
-        if (m_pInstance)
-            m_pInstance->SetData(TYPE_ANUB_REKHAN, DONE);
-    }
-
-    void JustReachedHome() override
-    {
-        if (m_pInstance)
-            m_pInstance->SetData(TYPE_ANUB_REKHAN, FAIL);
-
-        DoCastSpellIfCan(m_creature, SPELL_DOUBLE_ATTACK, CAST_TRIGGERED | CAST_AURA_NOT_PRESENT);
+        BossAI::KilledUnit(pVictim);
```

Comment:
I strongly suggest calling parent func before your stuff. This is more of a finnicky picky thing, but if blizzard uses generic scripting mechanisms, then events are prolly executed just like in EAI fashion after default behaviour.

Reply:
Understood